### ISSUE TITLE
Resource vanished while downloading new ghc

### ISSUE THREAD

**Issue by k-bx** [commercialhaskell/stack#854]
```
➜  stack git:(master) stack setup
ghc-7.10.2:    6.12 MiB / 140.12 MiB (  4.37%) downloaded...<socket: 18>: hGetBuf: resource vanished (Connection reset by peer)
➜  stack git:(master) stack setup
ghc-7.10.2:   31.77 MiB / 140.12 MiB ( 22.67%) downloaded...<socket: 18>: hGetBuf: resource vanished (Connection reset by peer)
➜  stack git:(master) stack setup
ghc-7.10.2:   45.75 MiB / 140.12 MiB ( 32.65%) downloaded...<socket: 18>: hGetBuf: resource vanished (Connection reset by peer)
```

I don't have the best internet in the world, I guess (but I downloaded bigger files also). Ideal would be to retry with continuing download, something similar to `wget -c`.

**Comment by markus1189** [comment]
Just an idea, but if stack uses `verifiedDownload` for the ghc download, I think we only have to include the exception from `hGetBuf` as another handler [here](https://github.com/commercialhaskell/stack/blob/master/src/Network/HTTP/Download/Verified.hs#L203).  Seems like `ResourceVanished` is an `IOError` currently we recover only from `HttpException`.

I'll do a pr later.

@k-bx is this reproducible, i.e. does this always happen if we disconnect from the internet while downloading?

**Comment by k-bx** [comment]
@markus1189 the problem isn't reproduced by disconnecting from internet. Disconnecting from internet makes download just freeze (without continuation on reconnect). In order to reproduce this bug we'd need to write a small server (instead of the one from where we download), and make it "stop" (exit or kill thread without closing socket) before we finish our download.

**Comment by markus1189** [comment]
Oh after rereading your first message I see it already did retry 3 times above.  That means we do not need to add more handlers,  one more idea would be to adjust the `drRetryPolicyDefault` to try a little harder?

**Comment by k-bx** [comment]
@markus1189 while I don't fully understand current logic, I believe your initial suggestion is correct. It didn't retry 3 times, I manually re-launched it 3 times, so either adding `IOError` handling to stack, or converting it to `HttpException` on http-library level seems to make sense.

**Comment by markus1189** [comment]
Oh okay then I misread the first message.  Actually the current logic is easy:

```haskell
drRetryPolicyDefault :: RetryPolicy
drRetryPolicyDefault = limitRetries 3 <> constantDelay onehundredMilliseconds
  where onehundredMilliseconds = 100000
```

This comes from https://hackage.haskell.org/package/retry.  In the default case, we do a maximum of three retries, waiting 100ms between each retry.  You can define more retry logic using a monoid based dsl from functions in https://hackage.haskell.org/package/retry-0.6/docs/Control-Retry.html#g:2

**Comment by k-bx** [comment]
@markus1189 so, points to address in this issue are:

1. make sure "resource vanished" gets also retried (just as `HttpException`)
2. continue download, similar to `wget -c`. GHC is quite a big beast, and re-downloading from scratch on network error is now what you'd want, so ideally would be to continue download instead of starting all over. This is quite a big task, I think, that's why I guess I'd put it lower on priority, but I think this should be done, as it would be quite frustrating for people without a good internet connection (it would be easier for them to use `wget -c` to manually download and install `ghc`).

**Comment by snoyberg** [comment]
Looks like this was addressed by PR #865 and we forgot to close, closing now.
